src/plugins/chart_expressions/expression_xy/public/components/xy_chart.tsx

```diff
@@ -163,27 +168,29 @@ export function XYChart({
   const chartTheme = chartsThemeService.useChartsTheme();
   const chartBaseTheme = chartsThemeService.useChartsBaseTheme();
   const darkMode = chartsThemeService.useDarkMode();
-  const filteredLayers = getFilteredLayers(layers, data);
-  const layersById = filteredLayers.reduce<Record<string, XYLayerConfigResult>>(
-    (hashMap, layer) => {
-      hashMap[layer.layerId] = layer;
+  const filteredLayers = getFilteredLayers(layers);
+  const layersById = filteredLayers.reduce<Record<string, CommonXYLayerConfigResult>>(
+    (hashMap, layer, index) => {
+      hashMap[index] = layer;
       return hashMap;
     },
     {}
   );
 
   const handleCursorUpdate = useActiveCursor(chartsActiveCursorService, chartRef, {
-    datatables: Object.values(data.tables),
+    datatables: layers.map(({ table }) => table),
   });
 
   if (filteredLayers.length === 0) {
     const icon: IconType = getIconForSeriesType(getDataLayers(layers)?.[0]?.seriesType || 'bar');
     return <EmptyPlaceholder className="xyChart__empty" icon={icon} />;
   }
 
+  const dataLayers: CommonXYDataLayerConfigResult[] = filteredLayers.filter(isDataLayer);
+
   // use formatting hint of first x axis column to format ticks
-  const xAxisColumn = data.tables[filteredLayers[0].layerId].columns.find(
-    ({ id }) => id === filteredLayers[0].xAccessor
+  const xAxisColumn = dataLayers[0]?.table.columns.find(
+    ({ id }) => isDataLayer(dataLayers[0]) && id === dataLayers[0].xAccessor
```

Comment:
Why we should use isDataLayer here? We've already used dataLayer[0] which should mean that is array of data layers.